src/connector.js

```diff
@@ -0,0 +1,150 @@
+'use strict';
+
+const net = require('net');
+const lookupAll = require('dns-lookup-all');
+
+class Connector {
+  constructor(options, multiSubnetFailover) {
+    this.options = options;
+    this.multiSubnetFailover = multiSubnetFailover;
+  }
+
+  execute(cb) {
+    if (net.isIP(this.options.host)) {
+      this.executeForIP(cb);
+    } else {
+      this.executeForHostname(cb);
+    }
+  }
+
+  executeForIP(cb) {
+    const socket = net.connect(this.options);
+
+    const onError = (err) => {
+      socket.removeListener('error', onError);
+      socket.removeListener('connect', onConnect);
+
+      socket.destroy();
+
+      cb(err);
+    };
+
+    const onConnect = () => {
+      socket.removeListener('error', onError);
+      socket.removeListener('connect', onConnect);
+
+      cb(null, socket);
+    };
+
+    socket.on('error', onError);
+    socket.on('connect', onConnect);
+  }
+
+  executeForHostname(cb) {
+    lookupAll(this.options.host, 4, (err, addresses) => {
```

Comment:
> we'll try connecting to each one of them twice.

Not sure what you mean there.

From https://nodejs.org/api/dns.html#dns_dns_lookup_hostname_options_callback:

_family <Number> - The record family. If present, must be the integer 4 or 6. If not provided, both IP v4 and v6 addresses are accepted._

That suggests, there is nothing special to do other than to pass on the user intent to lookupAll.